apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js

```diff
@@ -68,173 +199,105 @@ function processSubmission(req, res, callback) {
-    // Only close if the timer expired due to time limit, not for access end
-    if (!res.locals.assessment_instance_time_limit_expired) {
-      return res.redirect(req.originalUrl);
-    }
+    } else if (req.body.__action === 'timeLimitFinish') {
+      if (res.locals.assessment.type !== 'Exam') {
+        throw error.make(400, 'Only exams have a time limit');
+      }
+      // Only close if the timer expired due to time limit, not for access end
+      if (!res.locals.assessment_instance_time_limit_expired) {
+        return res.redirect(req.originalUrl);
+      }
 
-    const requireOpen = true;
-    const closeExam = true;
-    const overrideGradeRate = false;
-    assessment.gradeAssessmentInstance(
-      res.locals.assessment_instance.id,
-      res.locals.authn_user.user_id,
-      requireOpen,
-      closeExam,
-      overrideGradeRate,
-      function (err) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/assessment_instance/' +
-            res.locals.assessment_instance.id +
-            '?timeLimitExpired=true',
-        );
-      },
-    );
-  } else if (req.body.__action === 'attach_file') {
-    util.callbackify(studentInstanceQuestion.processFileUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'attach_text') {
-    util.callbackify(studentInstanceQuestion.processTextUpload)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'delete_file') {
-    util.callbackify(studentInstanceQuestion.processDeleteFile)(
-      req,
-      res,
-      function (err, variant_id) {
-        if (ERR(err, next)) return;
-        res.redirect(
-          res.locals.urlPrefix +
-            '/instance_question/' +
-            res.locals.instance_question.id +
-            '/?variant_id=' +
-            variant_id,
-        );
-      },
-    );
-  } else if (req.body.__action === 'report_issue') {
-    util.callbackify(studentInstanceQuestion.processIssue)(req, res, function (err, variant_id) {
-      if (ERR(err, next)) return;
+      const requireOpen = true;
+      const closeExam = true;
+      const overrideGradeRate = false;
+      await gradeAssessmentInstanceAsync(
+        res.locals.assessment_instance.id,
+        res.locals.authn_user.user_id,
+        requireOpen,
+        closeExam,
+        overrideGradeRate,
+        res.locals.client_fingerprint_id,
```

Comment:
👀 was this a bug before?

Reply:
It was indeed missing, so a client fingerprint would not be captured in a time-finish event that happened when the student was in an instance question.